project-lisa/source-code/client/src/api.ts

```diff
@@ -1,24 +1,242 @@
-/**
- * Exports of the lisa client.
- */
-export type LisaClient = {
-	load: (
-		path: string,
-		options: {
-			fs: unknown
-			auth?: unknown
-		},
-	) => Promise<Repository>
-}
+import raw from "isomorphic-git"
+import http from "isomorphic-git/http/web/index.js"
+import { createNodeishMemoryFs } from "@inlang-git/fs"
+import type { NodeishFilesystem } from "@inlang-git/fs"
+import { github } from "./github.js"
+import { withLazyFetching, transformRemote } from "./helpers.js"
 
 export type Repository = {
-	fs: any
-	commit: () => unknown
-	push: () => unknown
-	pull: () => unknown
-	currentBranch: () => unknown
-	changeBranch: () => unknown
-	listRemotes: () => unknown
-	log: () => unknown
-	status: () => any
+	fs: NodeishFilesystem
+	commit: (args: { author: any, message: string}) => Promise<Awaited<ReturnType<typeof raw.commit>> | undefined>
+	push: () => Promise<Awaited<ReturnType<typeof raw.push>> | undefined>
+	pull: (args: { author: any, fastForward: boolean, singleBranch: true }) => any
+	add: (args: { filepath: string }) =>  Promise<Awaited<ReturnType<typeof raw.add>>>
+	listRemotes: () => Promise<Awaited<ReturnType<typeof raw.listRemotes>> | undefined>
+	log: (args: { since: any}) =>  Promise<Awaited<ReturnType<typeof raw.log>>>
+	statusMatrix: (args: { filter: any }) => Promise<Awaited<ReturnType<typeof raw.statusMatrix>>>
+	status: (args: { filepath: string }) => Promise<Awaited<ReturnType<typeof raw.status>>>
+	mergeUpstream: (args: { branch: string }) => ReturnType<typeof github.request<"POST /repos/{owner}/{repo}/merge-upstream">>
+	isCollaborator: (args: { username: string }) => Promise<boolean>
+	getOrigin: () => Promise<string>
+	getCurrentBranch: () => Promise<string | undefined>
+	getMeta: () => Promise<{
+		name: string,
+		isPrivate: boolean,
+		isFork: boolean,
+		owner: { name?: string, email?: string },
+		parent:  {
+			url: string,
+			fullName: string
+		} | undefined
+	}>
+
+	// TODO: implement these before publishing api, but not used in badge or editor
+	// currentBranch: () => unknown
+	// changeBranch: () => unknown
+}
+
+export function open (url: string, args?: { nodeishFs?: NodeishFilesystem, workingDirectory?: string, corsProxy?: string, auth?: unknown }): Repository {
+	const rawFs = args?.nodeishFs ? args.nodeishFs : createNodeishMemoryFs()
```

Comment:
I would not create a memory fs under the hood. let users explicitly choose the fs they want to use. (also makes `args` not optional.)